Upgrade contribution policy and AI usage (#4989)
lima-vm/lima: website/content/en/docs/community/contributing.md

Reviewer: Probably too lengthy

We’re glad to have you here, your contributions make Lima better!

### Sending pull requests

Pull requests can be submitted to <https://github.com/lima-vm/lima/pulls>. Please ensure that you are familiar with the below policies before submitting pull requests.

> #### Talk first, code later:

Reviewer: ```suggestion
#### Talk first, code later
```

Before opening a pull request, open an issue first and explain your idea. Approval can be given as an approved comment, or by adding the `ready-to-work` label. It is okay to submit a pull request before the issue is approved, but keep in mind that unapproved work may not be merged and you risk wasting your effort.

**Exceptions (issue is not required first):**
- Pull requests from maintainers.
- Very small fixes _(for example typos, or a pull request touching fewer than 2 files with about 10 lines of change.)_
- Simple tool or dependency updates.

> #### One fix per pull request:

Reviewer: ```suggestion
#### One fix per pull request
```

Each pull request should fix one specific thing. Do not mix unrelated changes in one pull request. For large, ground-breaking work that needs many changes to test CI or integration, a draft pull request is okay first. After that, split the work into smaller pull requests that depend on each other.

It is highly suggested to add [tests](../../dev/testing/) for every non-trivial pull requests.
A test can be implemented as a unit test rather than an integration test when it is possible,
to avoid slowing the integration test CI.

> For tips on squashing commits and rebasing before submitting your pull request, see [Git Tips](../dev/git.md).

Reviewer 2: Add before pointing to `git` tips (the submitter needs to know that we expect PRs to be squashed).

```
All commits in a PR need to be squashed to a single commit before it can be merged. Rebase on the latest `master` branch in case GitHub shows that there are merge conflicts!
```

Reviewer 3: Single commit is still a (too?) hard rule, while it is better to have one commit per actual code change.

I think squashing a whole branch to one commit is an anti-pattern, but of course it should be cleaned up.

Reviewer: > All commits in a PR need to be squashed to a single commit before it can be merged.

Sometimes not, when a PR covers multiple topics.
Such a PR should be split to multiple PRs though.

Reviewer 3: There can be multiple commits on a single topic, as well.

Reviewer 2: I know, I know, but we don't need to be comprehensive. Maintainers can always allow exceptions. If we have a rule that a PR can only address one issue, then it makes sense to also require commits to be squashed.

We can add weasel-words, like "usually", to hint that exceptions are possible.

### AI Contribution Rules

Lima welcomes help from AI tools, but we only accept high-quality pull requests. These rules help keep review time useful and fair for maintainers.

#### Humans are responsible for all content

All content in your pull request whether written directly by you or generated with AI tools must be reviewed and edited by you. You are responsible for ensuring the description is clear, accurate, and free of errors. It is often discouraged to have long AI-generated text blocks for pull request description, as reviewers need clear explanations directly from the contributor. When reviewers leave review comments, reply yourself without relying on AI tools.

#### Legal sign-off (DCO)

AI tools cannot legally sign off code. Only the human submitting the code can add a `Signed-off-by` line. See [Developer Certificate of Origin](#developer-certificate-of-origin-dco) below.
If you use AI-generated code, you must:

- Read and check all generated code before submitting.
- Add your own `Signed-off-by` tag.
- Take full responsibility for the submitted code.

#### Mention AI usage

> If you used AI tools while preparing your pull request, disclose that in the pull request description using an `Assisted-by: AI_TOOL_NAME` trailer (see [Linux kernel coding assistants policy](https://docs.kernel.org/process/coding-assistants.html)). `Co-Authored-By` trailers added by AI tools are also acceptable and will not block a pull request from being merged.

Reviewer 3: Should this be "commit message trailer", or should it only go in the "pull request description"?

The kernel link says that it should be added while committing, not while preparing the PR

Reviewer 3: @unsuman

The kernel.org changelog has some examples of what it looks like there, if you want to review.

Author: I think we should only allow the trailers for PR description. For instance k8s docs specify this clearly that trailers should not be added to commits: https://kubernetes.io/blog/2026/06/26/open-source-maintainership-in-the-age-of-ai/#human-accountability

Reviewer 3: Right, but they also don't allow `Co-authored-by`

The PR comment doesn't actually go anywhere

Author: Hm, I think we can make an exception for this one because adding for a PR description shouldn't hurt😄

Reviewer: Kubernetes's policy is too strict. We are not bound to it.

**Exceptions (issue is not required first):**
- Pull requests from maintainers.
- Very small fixes _(for example typos, or a pull request touching fewer than 2 files with about 10 lines of change.)_
- Simple tool or dependency updates.

#### One fix per pull request:
Each pull request should fix one specific thing. Do not mix unrelated changes in one pull request. For large, ground-breaking work that needs many changes to test CI or integration, a draft pull request is okay first. After that, split the work into smaller pull requests that depend on each other.

It is highly suggested to add [tests](../../dev/testing/) for every non-trivial pull requests.
A test can be implemented as a unit test rather than an integration test when it is possible,
to avoid slowing the integration test CI.

For tips on squashing commits and rebasing before submitting your pull request, see [Git Tips](../dev/git.md).

### AI Contribution Rules

Lima welcomes help from AI tools, but we only accept high-quality pull requests. These rules help keep review time useful and fair for maintainers.

#### Humans should write the pull request description

> A human should write the pull request description in their own words. Avoid long AI-generated text blocks, as reviewers need clear explanations directly from the contributor. When reviewers leave review comments, reply yourself without relying on AI tools.

Reviewer: ```suggestion
A human should write the pull request description in their own words. It is often discouraged to have long AI-generated text blocks, as reviewers need clear explanations directly from the contributor. When reviewers leave review comments, reply yourself without relying on AI tools.
```

AI-generated description is sometimes acceptable

Reviewer 4: AI is great at writing, I don't write pull request messages or commit message now. AI write it and I review and refine it with the agent. Sometimes I edit manually but this typically add inconsistencies, typos and grammar issues. Same for replies, AI is very good at writing replies. This requirement is not helpful.

Maybe replace with requirement to review and edit code and text generated by AI - human is responsible for all content in a PR. No AI generated content should be added without review by human.

#### Legal sign-off (DCO)

AI tools cannot legally sign off code. Only the human submitting the code can add a `Signed-off-by` line. See [DCO](https://github.com/lima-vm/lima/blob/master/DCO.md).
If you use AI-generated code, you must:

- Read and check all generated code before submitting.
- Add your own `Signed-off-by` tag.
- Take full responsibility for the submitted code.

#### Mention AI usage

> If you used AI tools while preparing your pull request, disclose that in the pull request description using an `Assisted-by: AI_TOOL_NAME` trailer (see [Linux kernel coding assistants policy](https://docs.kernel.org/process/coding-assistants.html)).

Reviewer: I'd prefer to allow Claude's default trailer `Co-Authored-By: Claude * <noreply@anthropic.com>`.

Reviewer 4: This is also Cursor default when you let the use git commit. Very useful to allow it if this is OK with CNCF.

A bad side effect of using Co-Authored-By is that the agent becomes a contributor in the project, which contradicts the concept of human responsibility.

Reviewer 3: It is OK to list the AI as an author, even if it won't get any copyright by doing so. The DCO uses "Signed off"

Reviewer 3: I think introducing the Assisted was so that they could do data mining on it, like HTTP's User-Agent

"proper attribution helps track the evolving role of AI in the development process"

`Assisted-by: AGENT_NAME:MODEL_VERSION [TOOL1] [TOOL2]`

Author: > A bad side effect of using Co-Authored-By is that the agent becomes a contributor in the project

But as long as the trailer only lives in the PR description it should not show up in the contributor graph, right?

>I think introducing the Assisted was so that they could do data mining on it, like HTTP's User-Agent

Correct but I think it's upto the author what they want to specify. `Assisted-by: AI_TOOL_NAME` is preferred but `Co-Authored-By` trailers should also be acceptable and not block a PR.

Reviewer 3: We added some workarounds to filter out the bots, when listing the authors (another one was dependabot)

* https://github.com/lima-vm/lima/pull/4990

#### Enforcement

Maintainers may close pull requests that do not follow these rules.

> You can always open or update an issue first and come back with a compliant pull request.

Reviewer: But we may close such an issue and a compliant PR again?

```suggestion
You can leave a comment in the pull request when you think it was closed inappropriately.
```

We’re glad to have you here, your contributions make Lima better!

### Sending pull requests

Pull requests can be submitted to <https://github.com/lima-vm/lima/pulls>. Please ensure that you are familiar with the below policies before submitting pull requests.

#### Talk first, code later:

> Before opening a pull request, open an issue first and explain your idea. A maintainer must approve the idea before work starts. Approval can be written as an approved comment, or by adding the `ready-to-work` label. If a pull request is opened without a linked and approved issue, we will close it.

Reviewer: Too harsh.
See my comment above.

**Exceptions (issue is not required first):**
- Pull requests from maintainers.
- Very small fixes _(for example typos, or a pull request touching fewer than 2 files with about 10 lines of change.)_
- Simple tool or dependency updates.

#### One fix per pull request:
Each pull request should fix one specific thing. Do not mix unrelated changes in one pull request. For large, ground-breaking work that needs many changes to test CI or integration, a draft pull request is okay first. After that, split the work into smaller pull requests that depend on each other.

It is highly suggested to add [tests](../../dev/testing/) for every non-trivial pull requests.
A test can be implemented as a unit test rather than an integration test when it is possible,
to avoid slowing the integration test CI.

For tips on squashing commits and rebasing before submitting your pull request, see [Git Tips](../dev/git.md).

### AI Contribution Rules

Lima welcomes help from AI tools, but we only accept high-quality pull requests. These rules help keep review time useful and fair for maintainers.

> #### Humans must write the pull request description

Reviewer: must write → should write

They must review the AI-generated description anyway, though.

A human must write the pull request description and avoid long AI-generated text blocks, as reviewers need clear explanations directly from the contributor. When reviewers leave review comments, you must reply yourself without relying on AI tools.

#### Legal sign-off (DCO)

AI tools cannot legally sign off code which means it must never add a `Signed-off-by` line. Only the human submitting the code can sign off.
**If you use AI-generated code, you must:**

- Read and check all generated code before submitting.
- Add your own `Signed-off-by` tag.
- Take full responsibility for the submitted code.

#### Mention AI usage

If you used AI tools while preparing your pull request, disclose that in the pull request description.

**Not allowed:**

> - Listing an AI tool as a co-author.

Reviewer: We can allow this. Being an author in GitHub is different from being an author in academic papers.

Reviewer 3: They will not get any copyright, though. Even after being listed as author, at least not with current laws.

```
Copyright The Lima Authors.
```

* https://github.com/lima-vm/lima/pull/4990

- Co-signing commits with an AI tool.

> - Using `assisted-by`, `co-developed`, or similar commit trailers for AI tools.

Reviewer: This contradicts with PULL_REQUEST_TEMPLATE, which recommends assited-by trailer

Reviewer 3: And it is probably better to have the ads in the machine-readable trailer, than in the text...

```
    🤖 Generated with [Claude Code](https://claude.com/claude-code)
```

So using `Assisted-by` at least puts it all in the same place, and makes it easier to hide.

Author: I thought only to add the trailer in the PR desc but there are no downsides to adding it to commit messages as well

Reviewer 3: No downsides?

Author: My bad, we discussed this earlier in https://github.com/lima-vm/lima/issues/4982#issuecomment-4452493480 which I forgot. We should then explicitly mention that using the `Assisted-by` trailer is ok for PR desc and not for commit messages, right?

Reviewer 3: I don't think we need to add bad examples on what NOT to do, the commit message is about the commit.
